<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>practical3 – ESBS Bioinformatics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark-b53751a350365c71b6c909e95f209ed1.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-d7ba9c6adfd1938aed2b9fb196500684.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark-51e2c66856be53ee965980e392a92cdf.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": true,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "?",
    "H"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ESBS Bioinformatics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-a" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">1A</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-a">    
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical1.html">
 <span class="dropdown-text">1 - Exploring sequence database</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical2.html">
 <span class="dropdown-text">2 - Functional analysis of a gene list and sequence comparison</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical3.html">
 <span class="dropdown-text">3 - Similary searches</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical4.html">
 <span class="dropdown-text">4 - Analysis of Protein Families</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical5.html">
 <span class="dropdown-text">5 - Multiple Alignment &amp; Phylogeny</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-a-1" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">2A</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-a-1">    
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical1.html">
 <span class="dropdown-text">1 - Pre-processing of sequencing reads</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical2.html">
 <span class="dropdown-text">2 - Read mapping on a reference genome</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical3.html">
 <span class="dropdown-text">3 - De novo assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical4.html">
 <span class="dropdown-text">4 - Annotation of a complete prokaryotic genome</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical5.html">
 <span class="dropdown-text">5 - Analysis of eukaryotic sequences</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-a-2" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">3A</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-a-2">    
        <li>
    <a class="dropdown-item" href="../practicals-3A/practical2.html">
 <span class="dropdown-text">2 - Exome analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-3A/practical3.html">
 <span class="dropdown-text">3 - Personal genomics: variants</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-pharma" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Pharma</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-pharma">    
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical1.html">
 <span class="dropdown-text">1 - Pre-processing of sequencing reads</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical2.html">
 <span class="dropdown-text">2 - Exploring sequence database</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical3.html">
 <span class="dropdown-text">3 - Functional analysis of a gene list and sequence comparison</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical4.html">
 <span class="dropdown-text">4 - Similary searchese</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical5.html">
 <span class="dropdown-text">5 - Analysis of Protein Families</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-help" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Help</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-help">    
        <li>
    <a class="dropdown-item" href="../help/galaxy.html">
 <span class="dropdown-text">Galaxy Help</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../help/server_connection_guide.html">
 <span class="dropdown-text">Server Connection Guide</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/dorinemerlat/bioinformatics-en">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/dorinemerlat/bioinformatics-en/issues">
            Report a Bug
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#practical---similary-searches" id="toc-practical---similary-searches" class="nav-link active" data-scroll-target="#practical---similary-searches">Practical - Similary searches</a>
  <ul class="collapse">
  <li><a href="#search-for-homologous-proteins-using-blastp-ncbi-blast" id="toc-search-for-homologous-proteins-using-blastp-ncbi-blast" class="nav-link" data-scroll-target="#search-for-homologous-proteins-using-blastp-ncbi-blast">Search for homologous proteins using BLASTP (NCBI BLAST)</a>
  <ul class="collapse">
  <li><a href="#composition" id="toc-composition" class="nav-link" data-scroll-target="#composition">Composition</a></li>
  <li><a href="#databases" id="toc-databases" class="nav-link" data-scroll-target="#databases">Databases</a></li>
  <li><a href="#blastp-parameters" id="toc-blastp-parameters" class="nav-link" data-scroll-target="#blastp-parameters">BLASTP parameters</a></li>
  <li><a href="#perform-the-similarity-search-swissprot-e10-default-for-other-parameters" id="toc-perform-the-similarity-search-swissprot-e10-default-for-other-parameters" class="nav-link" data-scroll-target="#perform-the-similarity-search-swissprot-e10-default-for-other-parameters">Perform the similarity search (SwissProt, E≤10, default for other parameters)</a></li>
  <li><a href="#description" id="toc-description" class="nav-link" data-scroll-target="#description">Description</a></li>
  </ul></li>
  <li><a href="#localisation-of-a-transcript-on-a-genome-and-search-for-a-cds" id="toc-localisation-of-a-transcript-on-a-genome-and-search-for-a-cds" class="nav-link" data-scroll-target="#localisation-of-a-transcript-on-a-genome-and-search-for-a-cds">Localisation of a transcript on a genome and search for a CDS</a>
  <ul class="collapse">
  <li><a href="#localization-of-a-mrna-sequence-on-a-genome" id="toc-localization-of-a-mrna-sequence-on-a-genome" class="nav-link" data-scroll-target="#localization-of-a-mrna-sequence-on-a-genome">Localization of a mRNA sequence on a genome</a></li>
  </ul></li>
  <li><a href="#gene-identification-in-metagenomic-data" id="toc-gene-identification-in-metagenomic-data" class="nav-link" data-scroll-target="#gene-identification-in-metagenomic-data">Gene identification in metagenomic data</a>
  <ul class="collapse">
  <li><a href="#connection-to-the-server" id="toc-connection-to-the-server" class="nav-link" data-scroll-target="#connection-to-the-server">Connection to the Server</a></li>
  <li><a href="#exploring-data" id="toc-exploring-data" class="nav-link" data-scroll-target="#exploring-data">Exploring data</a></li>
  <li><a href="#make-database" id="toc-make-database" class="nav-link" data-scroll-target="#make-database">Make database</a></li>
  <li><a href="#blast" id="toc-blast" class="nav-link" data-scroll-target="#blast">BLAST</a></li>
  <li><a href="#playing-around-blast-output" id="toc-playing-around-blast-output" class="nav-link" data-scroll-target="#playing-around-blast-output">Playing around BLAST output</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"></header>




<section id="practical---similary-searches" class="level1">
<h1>Practical - Similary searches</h1>
<section id="search-for-homologous-proteins-using-blastp-ncbi-blast" class="level2">
<h2 class="anchored" data-anchor-id="search-for-homologous-proteins-using-blastp-ncbi-blast">Search for homologous proteins using BLASTP (NCBI BLAST)</h2>
<p>We want to find homologs of a bat protein sequence available on moodle.</p>
<section id="composition" class="level3">
<h3 class="anchored" data-anchor-id="composition">Composition</h3>
<p>What can you say about the composition of this protein sequence? What problem can be encountered in a similarity search with such a composition?</p>
</section>
<section id="databases" class="level3">
<h3 class="anchored" data-anchor-id="databases">Databases</h3>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 1</strong></span> Which databases are available for the search?</p>
</div>
<p>Select the SwissProt database.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 2</strong></span> Will the search be exhaustive?</p>
</div>
</section>
<section id="blastp-parameters" class="level3">
<h3 class="anchored" data-anchor-id="blastp-parameters">BLASTP parameters</h3>
<p>See Algorithm parameters, at the bottom of the page.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3</strong></span> How many sequences can you get?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4</strong></span> What is the Expect threshold to display sequences? Modify the Expect threshold to obtain results with E≤10.</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5</strong></span> What are the gap penalties? The word size?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 6</strong></span> How are regions of low complexity processed?</p>
</div>
</section>
<section id="perform-the-similarity-search-swissprot-e10-default-for-other-parameters" class="level3">
<h3 class="anchored" data-anchor-id="perform-the-similarity-search-swissprot-e10-default-for-other-parameters">Perform the similarity search (SwissProt, E≤10, default for other parameters)</h3>
<section id="graphic-summary" class="level4">
<h4 class="anchored" data-anchor-id="graphic-summary">Graphic summary</h4>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 7</strong></span> How many sequences do you obtain with E≤10?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 8</strong></span> What can be deduced from the graphical representation of BLAST results? There is no alignment on the C-terminal region. Why?</p>
</div>
</section>
</section>
<section id="description" class="level3">
<h3 class="anchored" data-anchor-id="description">Description</h3>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9</strong></span> Which families are detected?</p>
</div>
<section id="taxonomy" class="level4">
<h4 class="anchored" data-anchor-id="taxonomy">Taxonomy</h4>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 10</strong></span> In which organisms do you find hits?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 11</strong></span> In which species is the closest sequence found?</p>
</div>
</section>
<section id="alignments" class="level4">
<h4 class="anchored" data-anchor-id="alignments">Alignments</h4>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 12</strong></span> Identify a conserved motif in the superfamily.</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 13</strong></span> According to the alignments, are the detected sequences all homologous to our query?</p>
</div>
</section>
<section id="decrease-in-the-word-size" class="level4">
<h4 class="anchored" data-anchor-id="decrease-in-the-word-size">Decrease in the word size</h4>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 14</strong></span> What might be the consequences of a decrease in word size?</p>
</div>
<p>Repeat the search with the parameters used in question 4 but with a word size of 3 aa.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 15</strong></span> How many “hits” are you getting now?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 16</strong></span> In which taxon (taxa) are similar proteins found?</p>
</div>
</section>
</section>
</section>
<section id="localisation-of-a-transcript-on-a-genome-and-search-for-a-cds" class="level2">
<h2 class="anchored" data-anchor-id="localisation-of-a-transcript-on-a-genome-and-search-for-a-cds">Localisation of a transcript on a genome and search for a CDS</h2>
<p>A gene has been detected as being over-expressed in several patients with esophageal cancer. The sequence of the transcript is provided below.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span>human <span class="ex">mRNA</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">GTGTGGACACTCCTAGGTTAGAAAGTTTGGTATGTTGCTATACCTTTGCTTCTCCCACCT</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">TCCCCAATATCTAATATGTATTTCTCATTCTTAGAATAATCCAGAATGGCTACTCTGATC</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">TATGTTGATAAGGAAAATGGAGAACCAGGCACCCGTGTGGTTGCTAAGGATGGGCTGAAG</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">CTGGGGTCTGGACCTTCAATCAAAGCCTTAGATGGGAGATCTCAAGTTTCAACACCACGT</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">TTTGGCAAAACGTTCGATGCCCCACCAGCCTTACCTAAAGCTACTAGAAAGGCTTTGGGA</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">ACTGTCAACAGAGCTACAGAAAAGTCTGTAAAGACCAAGGGACCCCTCAAACAAAAACAG</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="ex">CCAAGCTTTTCTGCCAAAAAGATGACTGAGAAGACTGTTAAAGCAAAAAGCTCTGTTCCT</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">GCCTCAGATGATGCCTATCCAGAAATAGAAAAATTCTTTCCCTTCAATCCTCTAGACTTT</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">GAGAGTTTTGACCTGCCTGAAGAGCACCAGATTGCGCACCTCCCCTTGAGTGGAGTGCCT</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">CTCATGATCCTTGACGAGGAGAGAGAGCTTGAAAAGCTGTTTCAGCTGGGCCCCCCTTCA</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ex">CCTGTGAAGATGCCCTCTCCACCATGGGAATCCAATCTGTTGCAGTCTCCTTCAAGCATT</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="ex">CTGTCGACCCTGGATGTTGAATTGCCACCTGTTTGCTGTGACATAGATATTTAAATTTCT</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="ex">TAGTGCTTCAGAGTTTGTGTGTATTTGTATTAATAAAGCATTCTTTATCAGAAAAAAAAA</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="ex">AAAAAAA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="localization-of-a-mrna-sequence-on-a-genome" class="level3">
<h3 class="anchored" data-anchor-id="localization-of-a-mrna-sequence-on-a-genome">Localization of a mRNA sequence on a genome</h3>
<p>Localize the corresponding gene on the human genome with BLAST (BLAST Genomes section, at the bottom of the page). Use the GRCh38.p14 reference assembly.</p>
<section id="program" class="level4">
<h4 class="anchored" data-anchor-id="program">Program</h4>
<p>MEGABLAST is proposed by default.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 17</strong></span> What is the word size in MEGABLAST? Can we use MegaBLAST in our case or should we use BLASTN?</p>
</div>
</section>
<section id="analysis" class="level4">
<h4 class="anchored" data-anchor-id="analysis">Analysis</h4>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 18</strong></span> Comment obtained results (number of hits, conservation…)</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 19</strong></span> Give the localization of the transcribed region (chromosome, strand, start, end).</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 20</strong></span> How many exons are present in this gene? Can you give the precise boundaries of exons?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 21</strong></span> How can you explain the ranking provided by BLAST?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 22</strong></span> Which program can be used to improve your results? Compare results.</p>
</div>
</section>
<section id="localization-of-the-coding-sequence-cds" class="level4">
<h4 class="anchored" data-anchor-id="localization-of-the-coding-sequence-cds">Localization of the coding sequence (CDS)</h4>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 23</strong></span> How could we localize the coding region of this transcript?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 24</strong></span> Perform the similarity search in SwissProt.</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 25</strong></span> What are the boundaries and reading frame of the CDS?</p>
</div>
</section>
</section>
</section>
<section id="gene-identification-in-metagenomic-data" class="level2">
<h2 class="anchored" data-anchor-id="gene-identification-in-metagenomic-data">Gene identification in metagenomic data</h2>
<p>This exercise is based on content from <a href="https://blast-crash-course.readthedocs.io/en/latest/index.html">A Quick Guide to Getting Started with Shell and BLAST</a>.</p>
<p>The objective is to identify the existence of <strong>nitrogen fixation genes</strong>* in <strong>soil metagenome</strong>.</p>
<section id="connection-to-the-server" class="level3">
<h3 class="anchored" data-anchor-id="connection-to-the-server">Connection to the Server</h3>
<p>To learn how to connect to the server for practical work, refer to the <a href="../help/server_connection_guide.html">Server Connection Guide</a>.</p>
<p>Once you’re connected, start by creating a working directory for this session and navigating into it.</p>
<p><strong>mkdir</strong></p>
<p>The <code>mkdir</code> command stands for “make directory” and is used to create a new folder in the file system. You can create your directory like this:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="op">&lt;</span>directory_name<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>cd</strong></p>
<p>The cd command stands for “change directory”, allows you to move into a different folder within the file system. Use it as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> tp3</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> tp3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Help
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you need help with a command, display its help using: - <code>&lt;command-h</code> or <code>&lt;command--help</code> for short help messages. - <code>man &lt;command&gt;</code> to display the full manual of a command.</p>
</div>
</div>
</section>
<section id="exploring-data" class="level3">
<h3 class="anchored" data-anchor-id="exploring-data">Exploring data</h3>
<p>The materials used in this are located in the directory <code>/data/blast_tp/</code>.</p>
<p>The objective is to identify the existence of <strong>nitrogen fixation genes</strong>* in <strong>soil metagenome</strong>.</p>
<p><strong>ls</strong></p>
<p>The <code>ls</code> command is used to list the contents of a directory in Linux. It provides a way to see files and subdirectories within the specified directory.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="op">&lt;</span>directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>tree</strong></p>
<p>The <code>tree</code>command is used to display the contents of a directory in a hierarchical structure. It provides a visual representation of files and subdirectories within the specified directory.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> <span class="op">&lt;</span>directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 26</strong></span> What are in in <code>/data/blast_tp/</code> ? Use <code>ls</code>command. Explore the subdirectory with the <code>tree</code> command and map out the directory structure.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> /data/blast_tp/</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> /data/blast_tp/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Navigate to <code>blast_tp</code> directory thanks to <code>cd</code> command.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /data/blast_tp/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>This directory contains:</p>
<ul>
<li><code>nifh_db/</code>: a folder containing a nucleotide sequence file of 50 nifh genes (<code>nifh-ref.fa</code>), it’s the <strong>database</strong>.</li>
<li><code>query_gene/</code>: a folder contains metagenome files from three different Iowa crop soils (<code>corn.fa</code>, <code>soybean.fa</code>, and <code>prairie.fa</code>), there are the <strong>queries</strong>.</li>
</ul>
<p><strong>head</strong></p>
<p>The <code>head</code> command is used to display the first few lines of a file, which is helpful when you want to quickly preview its content without opening the entire file.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="op">&lt;</span>file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>By default, head shows the first 10 lines of the file. You can customize the number of lines displayed using the <code>-n</code> option. This option allows you to specify the exact number of lines to display.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 20 <span class="op">&lt;</span>file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command will display the first 20 lines of the file. You can adjust the number of lines according to your needs.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 27</strong></span> How does sequence fasta file look like? Use the command <code>head</code> to display the begining of each file.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You can display the contents of each file individually:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> metags/corn.fa </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> metags/prairie.fa </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> metags/soybean.fa</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> nifh_db/nifh-ref.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Or use <code>*</code> to apply the head command to all files at once:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="pp">*</span>/<span class="pp">*</span>.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The <code>*</code> is a wildcard that matches multiple files. In this case: <code>*/*.fa</code> expands to <strong>all .fa files located inside any subdirectory</strong>.</p>
<p>This approach is useful when working with multiple files in different directories, avoiding the need to list them manually.</p>
</div>
</div>
</div>
<p><strong>grep</strong></p>
<p><code>grep</code>, short for “global regular expression print”, is a command used for searching and matching text patterns in files contained in the regular expressions.</p>
<p>Display the lines with a pattern:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&lt;pattern&gt;"</span> <span class="op">&lt;</span>file_path<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command searches through the specified file and outputs the lines that contain the given pattern.</p>
<p>Count the lines with a pattern:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&lt;pattern&gt;"</span> <span class="op">&lt;</span>file_path<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The -c option tells grep to count the number of lines that match the specified pattern rather than displaying them. This is useful for quickly determining how many occurrences of the pattern exist in the file.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 28</strong></span> How many sequences are there in theses fasta files?</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You can count for each file individually:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> metags/corn.fa </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> metags/prairie.fa </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> metags/soybean.fa</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> nifh_db/nifh-ref.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Or use <code>*</code> to apply the grep command to all files at once:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> <span class="pp">*</span>/<span class="pp">*</span>.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The <code>*</code> is a wildcard that matches multiple files. In this case: <code>*/*.fa</code> expands to <strong>all .fa files located inside any subdirectory</strong>.</p>
<p>This approach is useful when working with multiple files in different directories, avoiding the need to list them manually.</p>
</div>
</div>
</div>
</section>
<section id="make-database" class="level3">
<h3 class="anchored" data-anchor-id="make-database">Make database</h3>
<p>We want to find the existence of nifh genes in soil metagenome samples. Making a database for nifh genes is a smart way to index the gene sequences so that it can search and compare metagenome sequences (query) with nifh gene sequences much faster.</p>
<p><strong>makeblastdb</strong></p>
<p>The <code>makeblastdb</code> command is used to create a BLAST database from your sequence data, which serves as the reference for comparing queries. Creating a database for genes is an efficient way to index gene sequences, allowing for much faster searches and comparisons against the database.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">makeblastdb</span> <span class="at">-in</span> <span class="op">&lt;</span>sequence_file<span class="op">&gt;</span> -dbtype <span class="op">&lt;</span>type<span class="op">&gt;</span> -out <span class="op">&lt;</span>name<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Explanation of the command options:</p>
<ul>
<li><code>-in &lt;sequence_file</code>&gt;: Specifies the input file containing the sequences you want to include in the database. This file should be in FASTA or other supported formats.</li>
<li><code>-dbtype &lt;type&gt;</code>: Indicates the type of sequences in the input file. Use <code>nucl</code> for nucleotide sequences and <code>prot</code> for protein sequences.</li>
<li><code>-out &lt;name&gt;</code>: Defines the name for the output database files. The program will create several files with this base name, which will be used for subsequent BLAST searches.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">makeblastdb</span> <span class="at">-in</span> <span class="op">&lt;</span>sequence_file<span class="op">&gt;</span> -dbtype <span class="op">&lt;</span>type<span class="op">&gt;</span> -out <span class="op">&lt;</span>name<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Navigate back to the directory of the current practical and construct the BLAST database.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/td3/</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="ex">makeblastdb</span> <span class="at">-in</span> /data/blast_tp/nifh_db/nifh-ref.fa <span class="at">-dbtype</span> nucl <span class="at">-out</span> nifh-db.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>Note: The tilde <code>~</code> is a shortcut for your home directory <code>/home/&lt;username&gt;</code>.</em></p>
</div>
</div>
</div>
<p>Look the files generared. Use <code>ls</code> command.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="blast" class="level3">
<h3 class="anchored" data-anchor-id="blast">BLAST</h3>
<p>Now, we can perform a BLAST search of the gene sequence against this database. BLAST offers different algorithms depending on the type of query and the database:</p>
<ul>
<li><strong>blastn</strong>: Used for nucleotide sequence queries against a nucleotide database.</li>
<li><strong>blastp</strong>: Used for protein sequence queries against a protein database.</li>
<li><strong>blastx</strong>: Used for translating a nucleotide query to protein and comparing it against a protein database.</li>
<li><strong>tblastn</strong>: Used for protein queries against a translated nucleotide database.</li>
<li><strong>tblastx</strong>: Used for translating both the nucleotide query and the database sequences, comparing them in protein space.</li>
</ul>
<p>To see the list of parameters of this command use:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-help</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For this practical, we need of theses options:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-query</span> <span class="op">&lt;</span>query_file<span class="op">&gt;</span> -db <span class="op">&lt;</span>db_file<span class="op">&gt;</span> -out <span class="op">&lt;</span>name_of_output<span class="op">&gt;</span> -outfmt <span class="op">&lt;</span>format<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Explanation of options:</p>
<ul>
<li><code>query &lt;query_file&gt;</code>: Specifies the input file with the sequence to search.</li>
<li><code>db &lt;db_file&gt;</code>: Specifies the BLAST database to search against.</li>
<li><code>out &lt;name_of_output&gt;</code>: Defines the output file for the results.</li>
<li><code>outfmt &lt;format&gt;</code>: Sets the output format to tabular for easier parsing. Format <code>6</code> provides tabular output, which includes information like query sequence, subject sequence, percent identity, alignment length, and e-value. It is commonly used for further parsing and analysis of the results. We can have a good understanding of <code>-outfmt 6</code> format from this <a href="https://sites.google.com/site/wiki4metagenomics/tools/blast/blastn-output-format-6">webpage</a>.</li>
</ul>
<p>Launch BLAST for each query file.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-query</span> /data/blast_tp/metags/corn.fa <span class="at">-db</span> nifh-db.fa <span class="at">-out</span> corn.out <span class="at">-outfmt</span> 6</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-query</span> /data/blast_tp/metags/prairie.fa <span class="at">-db</span> nifh-db.fa <span class="at">-out</span> prairie.out <span class="at">-outfmt</span> 6</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-query</span> /data/blast_tp/metags/soybean.fa <span class="at">-db</span> nifh-db.fa <span class="at">-out</span> soybean.out <span class="at">-outfmt</span> 6</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="playing-around-blast-output" class="level3">
<h3 class="anchored" data-anchor-id="playing-around-blast-output">Playing around BLAST output</h3>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 29</strong></span> How many results do you get for each metagenome? Use the <code>wc -l</code>.</p>
</div>
<p><strong>wc</strong></p>
<p>The <code>wc</code> (word count) command in Unix/Linux is a utility used to count lines, words, and bytes in a file or input stream.</p>
<p>Here are its main options:</p>
<ul>
<li><code>-l</code>: Counts the number of lines.</li>
<li><code>-w</code>: Counts the number of words.</li>
<li><code>-c</code>: Counts the number of bytes.</li>
<li><code>-m</code>: Counts the number of characters.</li>
<li><code>-L</code>: Reports the length of the longest line.</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="op">&lt;</span>file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> <span class="pp">*</span>.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The wildcard <code>*</code> matches all files ending with <code>.out</code>.</p>
</div>
</div>
</div>
<p>To understand the meaning of each column in the BLAST output (format 6), refer to this <a href="https://sites.google.com/site/wiki4metagenomics/tools/blast/blastn-output-format-6">webpage</a>.</p>
<p><strong>more</strong></p>
<p>You can display the file with the command <code>more</code>.</p>
<p>The <code>more</code> command is used to display the contents of a file one screen at a time. This is useful for viewing large files like FASTQ files, where the data is too big to fit on a single screen.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> <span class="op">&lt;</span>path_file_to_display<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 30</strong></span> What is the range of percentage identity for the results of BLAST of soybean metagenome?? Similarly, what is the range of e-values? Use <code>more</code>command.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> soybean.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p><strong>sort</strong></p>
<p>The <code>sort</code> command is used to sort the lines of a file or input stream based on specified criteria. It can sort in ascending or descending order and handle various types of data.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">file</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Common Options:</p>
<ul>
<li><code>-n</code>: Sort numerically.</li>
<li><code>-r</code>: Sort in reverse order.</li>
<li><code>-k &lt;key&gt;</code>: Specifies the key (column) to sort by.</li>
<li><code>-u</code>: Remove duplicate lines.</li>
</ul>
<p><strong>cut</strong></p>
<p>The <code>cut</code> command is used to extract specific sections or columns from each line of a file or input stream. It is particularly useful for parsing delimited files, such as CSV or tab-separated files.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">file</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Common Options:</p>
<ul>
<li><code>-d &lt;delimiter&gt;</code>: Specifies the delimiter character (default is tab).</li>
<li><code>-f &lt;fields&gt;</code>: Specifies the fields to extract (e.g., -f 1,3 to extract the first and third fields).</li>
</ul>
<p><strong>uniq</strong></p>
<p>The <code>uniq</code> command is used to remove (or keep) consecutive duplicate lines from a sorted file or input stream. It is often used in conjunction with sort to ensure that the input is sorted.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">uniq</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">file</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Common Options:</p>
<ul>
<li><code>-c</code>: Prefix lines with the number of occurrences.</li>
<li><code>-d</code>: Only print duplicate lines.</li>
<li><code>-u</code>: Only print unique lines.</li>
</ul>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 31</strong></span> How many uniq nifh genes were hit by sequences from soybean metagenome?</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are several ways to do this:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 soybean.out <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 soybean.out <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 32</strong></span> How many uniq sequences from prairie metagenome can hit nifh genes?</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are several ways to do this:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 prairie.out <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="kw">|</span><span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 prairie.out <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 33</strong></span> How to get the <code>sseqid</code>, <code>sstart</code>, and <code>send</code> column (column 2, 9, 10) from the soybean blast output file?</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2,9,10 soybean.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p><strong>grep</strong></p>
<p>The <code>grep</code> command is used to search for patterns within files or input streams. It stands for “global regular expression print” and is highly versatile for text processing.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">pattern</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">file</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Common Options:</p>
<ul>
<li><code>-i</code>: Ignore case.</li>
<li><code>-v</code>: Invert match (show lines that do not match the pattern).</li>
<li><code>-c</code>: Count the number of matching lines.</li>
<li><code>-n</code>: Show line numbers.</li>
<li><code>-r</code>: Recursively search directories.</li>
</ul>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 34</strong></span> How many times was gene <code>gi|909637271|emb|LN713523.1|</code> hit by sequences from corn metagenome sample:</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"gi|909637271|emb|LN713523.1|"</span> corn.out <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script>
    var custom_title = document.querySelectorAll('.custom .theorem-title');
    
    for (let i = 0; i < custom_title.length; i++ ) {
       var mod_name = custom_title[i].innerHTML;
       custom_title[i].innerHTML = mod_name.replace("Exercise", "Question");
    };
    </script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Copyright 2024, Dorine Merlat. Built with Quarto.</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>