# Practical – Analysis of Protein Families

## Characterization of a Sequence of Unknown Function

```bash
> Hypothetical protein [Candidatus Poribacteria sp. WGA-A3]
MVDFNNTLSSTASVQVAAHWDPIENTANNLHRSTEEKFNRDKAQWQEPVEMTWEQWLEVFNPGPAHPLKN
YSTADFQVFLPPSTVNVADVWDLDTEGILPFLRQFHPSATMKLPRYGSIPSDQKDAKACLRALSPEYADI
VFRIHARFTLDASIDAYFMPAQFAGHLIINRNSRTIHQWTLSLPNRNSNVDIGAFRSHDIGFVPRMELCS
VSETQPESIVWEAAITAEEADKKFQNSLYKFAEIEWTPIEEAVELAKASNRSIHAVLLFGVLDDESC
```

Search for the organism **Candidatus Poribacteria** in the [NCBI Taxonomy database](https://www.ncbi.nlm.nih.gov/taxonomy). 

::: {#exr-1 .custom}
What can you say about the organism taxonomy?
:::

 Search for homologs of the Poribacteria sequence (available here or on moodle) using [NCBI BLASTP](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE=Proteins):

- In the **RefSeq_protein** database (not in RefSeq select!!!),
- With word size of **3**.
- With a maximum of **1000 sequences** and **E≤10** (other parameters: default).

::: {#exr-1 .custom}
How many hits did you get with E ≤ 10?
:::

::: {#exr-1 .custom}
According to BLASTP alignments, propose 2 patterns conserved in the protein family.
:::


::: {#exr-1 .custom}
In which taxonomic group(s) do we find hits of the protein family? Propose a function and an evolutionary origin for the gene encoding the Poribacteria protein.
:::

:::

::: {#exr-1 .custom}
Look at the alignment with human selenoprotein N **NP_996809.1**. Do the alignments cover the C-terminal part of the query sequence and animal selenoprotein N? In metazoans, which amino acid "matches" the stop codon of the Poribacteria sequence? Based on this analysis, what hypothesis can be proposed? How could this hypothesis be tested?
:::

## Connection to the Server

To learn how to connect to the server for practical work, refer to the [Server Connection Guide](../help/server_connection_guide.qmd).

Once you're connected, start by creating a working directory for this session and navigating into it.

**mkdir**

The `mkdir` command stands for "make directory" and is used to create a new folder in the file system. You can create your directory like this:

```bash
mkdir <directory_name>
```

**cd**

The cd command stands for "change directory", allows you to move into a different folder within the file system. Use it as follows:

```bash
cd <directory>
```

::: {.callout-tip collapse="true"}
# Solution
```bash
mkdir tp4
cd tp4
```
:::

::: {.callout-note}
## Help
If you need help with a command, display its help using:
 - `<command-h` or `<command--help` for short help messages.
 - `man <command>` to display the full manual of a command.
:::

## Comparison Protein/Genome Using Command-Line BLAST

Retrieve the **Human Protein Sequence of the Selenoprotein N** (Accession Number: **Q9NZV5**) using the command GET of the UniProtKB API.

Documentation of UniProtKB API: [UniProtKB API Documentation](https://www.ebi.ac.uk/proteins/api/doc/#!/proteins/search)

::: {.callout-tip collapse="true"}
# Solution
```bash
curl -X GET --header 'Accept:text/x-fasta' -o Q9NZV5.fasta 'https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=100&accession=Q9NZV5'
```
:::

You can open the file with the `cat` command:
```bash
cat Q9NZV5.fasta
```

The sequence of Poribacteria genome is available at:  `/data/poribacteria/poribacteria_genomes.fasta`


Create the BLAST database with the `makeblastdb` command.

To see available options: 
```bash
makeblastdb –help
```

::: {.callout-tip collapse="true"}
# Solution
```bash
makeblastdb -in /data/poribacteria/poribacteria_genomes.fasta -dbtype nucl -out poribacteriaDb
```
:::

Perform a Similarity Search using the human selenoprotein N **Q9NZV5.fasta** as a query in the **PoribacteriaDb** database using the **tblastN** command.

To see available options: 
```bash
tblastn –help
```

::: {.callout-tip collapse="true"}
# Solution
```bash
tblastn -db poribacteriaDb -query Q9NZV5.fasta -out selN.tblastn
```
:::

Analysis of the TblastN Results. 

Display the TblastN output:

```bash
more selN.tblastn
```

::: {#exr-1 .custom}
Can we extend the gene sequence in 3’?
:::

::: {#exr-1 .custom}
What do you learn from this search about the evolution of this gene in available Poribacteria genomes?
:::

::: {#exr-1 .custom}
We obtain several alignments for the sequence **VYCT01000218**. Why?
:::

For that, use `grep`command with `-A` option. The `-A` option tells grep to display not only the matching line but also the subsequent *n* lines after each match.

::: {.callout-tip collapse="true"}
# Solution
```bash
grep -A 90 '>VYCT01000218' selN.tblastn
```
:::

## Analysis of a Protein Family

A protein sequence (available on the Moodle platform) has been automatically predicted during the annotation of the fungus *Malassezia globosa* genome.

### Similarity Search

Perform a similarity search using BLAST at EBI in the SWISSPROT database (https://www.ebi.ac.uk/jdispatcher/sss/ncbiblast) (default parameters).

::: {#exr-1 .custom}
How many sequences do you obtain (with E<10)? Why so few sequences are detected according to you?  
:::

Display results with Tool output.

::: {#exr-1 .custom}
Find 3 conserved motifs.  
:::

::: {#exr-1 .custom}
Is the *Malassezia* sequence homologous to: Q84424, P05469, P0C995, O13881?  
:::

Display results with Visual output.

::: {#exr-1 .custom}
For each group of homologs, specify which region(s) (N-terminus, central region, C-terminus, almost whole sequence...) of the query sequence and of the database sequences have been aligned by BLAST.
:::

::: {#exr-1 .custom}
Why can this information be important before constructing a global multiple alignment?  
:::

Display the Summary table. Select all sequences detected by BLAST except:  

- CLR2_SCHPO (O13881) and YKP3_KLULA (P05469): non-homologous sequences  
- YL226_MIMIV (Q5UQC0) and MCE_MIMIV (Q5UQX1): too few motifs.  
  
Save sequences in Fasta format and add the query sequence in the file. The final file is also available on Moodle (`distant_homologs.txt`).

Keep the sequences for the next practical. 