# Practical - Pre-processing of Sequencing Reads

To practise quality control and pre-processing of raw NGS data, we will use small datasets. The tools are available in Galaxy, a platform for omics data analysis.

First step: log on to https://usegalaxy.fr

then, to ensure you have sufficient resources, go to the url:

https://usegalaxy.fr/join-training/td1genomics/

You can then return to galaxy.

## Illumina SE

You can rename your history (right panel).

The 1st data set concerns a metagenomic study of the microbiome of the oral cavity of a snake. After extraction and fragmentation of the sample DNA, part of the 16S rRNA gene was amplified and sequenced (Illumina Single End) in order to identify the bacteria present. Due to time constraints, only part of the data will be analysed.

### Raw data: FASTQ file

1. Import the FASTQ file from the zenodo server:

Copy the address of the link:
https://zenodo.org/record/3977236/files/female_oral2.fastq-4143.gz

In Galaxy, use the Upload Data button (top left) then Paste/Fetch data, copy the link and launch (start).

2. Look at the file (eye icon).

::: {#exr-1 .custom} 
What are the characteristics of the FASTQ format?
:::

::: {#exr-1 .custom}
Given that the scores are encoded using the SANGER/Illumina 1.8+ system, what is the score value for most of the bases at the start of the 1st read? Is this satisfactory?
:::

### Quality control using FASTQ

Open the FASTQ Quality Control toolbox (left panel) and choose FastQC.

**FastQC**

Explore the quality of reads with FastQC (default settings). See the FastQC website for help and examples of analysis.

::: {#exr-1 .custom}
Per Base Sequence Quality: How long are the reads? How does the quality evolve along the reads?
:::

::: {#exr-1 .custom}
Per sequence quality scores: What can be deduced from this graph?
:::

::: {#exr-1 .custom}
Per base sequence content: In the case of genome sequencing (without specific amplification), what should this graph look like? Is the result obtained normal in our case?
:::

::: {#exr-1 .custom}
Adapter content: Which adapter is detected?
:::

### Read processing using Trimmomatic

1. Use Trimmomatic (FASTQ Quality control section) to improve the quality of reads:

- Perform initial IllUMINACLIP step? Yes, select Nextera
- Sliding window trimming (proposed by default)
- Drop reads below a specific length (MINLEN): 20

::: {#exr-1 .custom}
Perform FastQC analysis again on the processed reads and compare the results to the raw data. Is the result acceptable?
:::

::: {#exr-1 .custom}
What would be the next step to identify the microbial species present in the snake's oral cavity from these reads?
:::

## PacBio Hifi

This dataset contains 50,000 PacBio HiFi reads from the sequencing of a human genome (PacBio - GIAB sample HG002).

### Import the file into Galaxy:

https://zenodo.org/api/files/ff9aa6e3-3d69-451f-9798-7ea69b475989/m64011_190830_220126.Q20.subsample.fastq.gz

2. Perform quality control with Nanoplot (available in Nanopore). This tool can only be used for long reads (PacBio or Nanopore):

- Options for customizing the plot created:
- Specify the bivariate format of the plots: dot
- Show the N50 mark in the read length histogram: Yes

### Analysis:

::: {#exr-1 .custom}
What is the average length of reads? Significant variations in read size are observed in the PacBio data, unlike the Illumina data. Why?
:::

::: {#exr-1 .custom}
What does N5O represent in this case?
:::

::: {#exr-1 .custom}
How useful can the Yield by length chart be?
:::

::: {#exr-1 .custom}
Read length vs Average read qualiy plot using dots: No read is observed with Q<20. Why? Is there a link between read size and quality?
:::

## NanoPore

The last set corresponds to nanopore reads obtained from a plant genome. We will use PycoQC for quality control. This only works with Nanopores readings; it requires specific information provided by nanopore basecallers such as Guppy.

### Import the files into Galaxy:

https://zenodo.org/api/files/ff9aa6e3-3d69-451f-9798-7ea69b475989/nanopore_basecalled-guppy.fastq.gz
https://zenodo.org/api/files/ff9aa6e3-3d69-451f-9798-7ea69b475989/sequencing_summary.txt

### Perform quality control with PycoQC

::: {#exr-1 .custom}
How many reads are available? How many bases does this represent?
:::

::: {#exr-1 .custom}
How do you explain the discrepancy between median reading size and N50?
:::

::: {#exr-1 .custom}
Basecalled reads length: what can you say about the distribution of lengths?
:::

::: {#exr-1 .custom}
Basecalled reads quality: What range do most quality scores fall into?
:::