<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>practical2 – ESBS Bioinformatics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark-b53751a350365c71b6c909e95f209ed1.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-d7ba9c6adfd1938aed2b9fb196500684.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark-51e2c66856be53ee965980e392a92cdf.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": true,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "?",
    "H"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ESBS Bioinformatics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-a" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">1A</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-a">    
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical1.html">
 <span class="dropdown-text">1 - Exploring sequence database</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical2.html">
 <span class="dropdown-text">2 - Functional analysis of a gene list and sequence comparison</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical3.html">
 <span class="dropdown-text">3 - Similary searches</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical4.html">
 <span class="dropdown-text">4 - Analysis of Protein Families</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-1A/practical5.html">
 <span class="dropdown-text">5 - Multiple Alignment &amp; Phylogeny</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-a-1" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">2A</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-a-1">    
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical1.html">
 <span class="dropdown-text">1 - Pre-processing of sequencing reads</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical2.html">
 <span class="dropdown-text">2 - Read mapping on a reference genome</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical3.html">
 <span class="dropdown-text">3 - De novo assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical4.html">
 <span class="dropdown-text">4 - Annotation of a complete prokaryotic genome</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-2A/practical5.html">
 <span class="dropdown-text">5 - Analysis of eukaryotic sequences</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-a-2" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">3A</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-a-2">    
        <li>
    <a class="dropdown-item" href="../practicals-3A/practical2.html">
 <span class="dropdown-text">2 - Exome analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-3A/practical3.html">
 <span class="dropdown-text">3 - Personal genomics: variants</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-pharma" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Pharma</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-pharma">    
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical1.html">
 <span class="dropdown-text">1 - Pre-processing of sequencing reads</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical2.html">
 <span class="dropdown-text">2 - Exploring sequence database</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical3.html">
 <span class="dropdown-text">3 - Functional analysis of a gene list and sequence comparison</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical4.html">
 <span class="dropdown-text">4 - Similary searchese</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals-pharma/practical5.html">
 <span class="dropdown-text">5 - Analysis of Protein Families</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-help" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Help</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-help">    
        <li>
    <a class="dropdown-item" href="../help/galaxy.html">
 <span class="dropdown-text">Galaxy Help</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../help/server_connection_guide.html">
 <span class="dropdown-text">Server Connection Guide</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/dorinemerlat/bioinformatics-en">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/dorinemerlat/bioinformatics-en/issues">
            Report a Bug
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#practical---exome-analysis-snv-calling" id="toc-practical---exome-analysis-snv-calling" class="nav-link active" data-scroll-target="#practical---exome-analysis-snv-calling">Practical - Exome analysis (SNV calling)</a>
  <ul class="collapse">
  <li><a href="#connection-to-the-server" id="toc-connection-to-the-server" class="nav-link" data-scroll-target="#connection-to-the-server">Connection to the server</a></li>
  <li><a href="#raw-sequence-data-exome-of-chromosome-22" id="toc-raw-sequence-data-exome-of-chromosome-22" class="nav-link" data-scroll-target="#raw-sequence-data-exome-of-chromosome-22">Raw sequence data: exome of chromosome 22</a></li>
  <li><a href="#fastq---quality-control" id="toc-fastq---quality-control" class="nav-link" data-scroll-target="#fastq---quality-control">FASTQ - Quality control</a></li>
  <li><a href="#alignement---mapping-to-the-reference-genome-fastq-to-bam" id="toc-alignement---mapping-to-the-reference-genome-fastq-to-bam" class="nav-link" data-scroll-target="#alignement---mapping-to-the-reference-genome-fastq-to-bam">Alignement - Mapping to the reference genome (FASTQ to BAM)</a>
  <ul class="collapse">
  <li><a href="#indexing-reference-file" id="toc-indexing-reference-file" class="nav-link" data-scroll-target="#indexing-reference-file">Indexing reference file</a></li>
  <li><a href="#alignment-using-bwa" id="toc-alignment-using-bwa" class="nav-link" data-scroll-target="#alignment-using-bwa">Alignment using BWA</a></li>
  <li><a href="#visualization-of-the-bam-file" id="toc-visualization-of-the-bam-file" class="nav-link" data-scroll-target="#visualization-of-the-bam-file">Visualization of the BAM file</a></li>
  </ul></li>
  <li><a href="#calling-single-nucleotide-variations-snvs-pileup-interval-and-bed-files" id="toc-calling-single-nucleotide-variations-snvs-pileup-interval-and-bed-files" class="nav-link" data-scroll-target="#calling-single-nucleotide-variations-snvs-pileup-interval-and-bed-files">Calling single nucleotide variations (SNVs) – pileup, interval and BED files</a>
  <ul class="collapse">
  <li><a href="#generation-of-a-pileup-file" id="toc-generation-of-a-pileup-file" class="nav-link" data-scroll-target="#generation-of-a-pileup-file">Generation of a Pileup file</a></li>
  <li><a href="#filtering-pileup-to-get-a-list-of-snvs" id="toc-filtering-pileup-to-get-a-list-of-snvs" class="nav-link" data-scroll-target="#filtering-pileup-to-get-a-list-of-snvs">Filtering pileup to get a list of SNVs</a></li>
  </ul></li>
  <li><a href="#annotation-of-variants" id="toc-annotation-of-variants" class="nav-link" data-scroll-target="#annotation-of-variants">Annotation of variants</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"></header>




<section id="practical---exome-analysis-snv-calling" class="level1">
<h1>Practical - Exome analysis (SNV calling)</h1>
<section id="connection-to-the-server" class="level2">
<h2 class="anchored" data-anchor-id="connection-to-the-server">Connection to the server</h2>
<p>To learn how to connect to the server for practical work, refer to the <a href="../help/server_connection_guide.html">Server Connection Guide</a>.</p>
<p>Once you’re connected, start by creating a working directory for this session and navigate into it.</p>
<p><strong>mkdir</strong></p>
<p>The mkdir command stands for “make directory”, is used to create a new folder in the file system. You can create your directory like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="op">&lt;</span>directory_name<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>cd</strong></p>
<p>The cd command stands for “change directory”, allows you to move into a different folder within the file system. Use it as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> TdExome</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> TdExome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Help
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you need help with a command, display its help using:</p>
<ul>
<li><code>&lt;command-h</code> or <code>&lt;command--help</code> for short help messages.</li>
<li><code>man &lt;command&gt;</code> to display the full manual of a command.</li>
</ul>
</div>
</div>
</section>
<section id="raw-sequence-data-exome-of-chromosome-22" class="level2">
<h2 class="anchored" data-anchor-id="raw-sequence-data-exome-of-chromosome-22">Raw sequence data: exome of chromosome 22</h2>
<p>Reads have been obtained using Illumina sequencing (Single-end library) from the exome of chromosome 22 of a single human individual, a Caucasian female from US. Her family was one of the original trios in Hapmap project and she was part of the 1000 Genomes project.</p>
<p>Create an alias of the file with reads in your directory for easy access. For that, use the command <code>ln</code>, which create links between files. Links are pointers to files, allowing multiple references (aliases) to the same file. This way, you don’t need to duplicate large files, saving storage space and making file management easier.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> <span class="op">&lt;</span>path_to_original_file<span class="op">&gt;</span> <span class="op">&lt;</span>path_to_link_file<span class="op">&gt;</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The option <code>-s</code>stands for “symbolic link.” A symbolic link is like a shortcut that points to the original file, enabling easy access. Unlike hard links, symbolic links can span different file systems, and if the original file is moved, the symbolic link can still point to it if it is relocated in the same system.</p>
<p>Create an alias for the reads (original path:<code>/data/human/reads/chrom22.fastq</code>) inside your directory.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /data/human/reads/chrom22.fastq .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The dot (<code>.</code>) refers to the current working directory. When you execute the command, this tells the system to create the symbolic link (alias) for chrom22.fastq in the current directory where you are running the command.</p>
</div>
</div>
</div>
<p>Create a directory for the genome reference of the human (original path: <code>/data/human/reference/human_genome.fasta</code>) and create an alias for the file.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> reference</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /data/human/reference/human_genome.fasta reference</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Display data with the command <code>more</code>.</p>
<p>The more command is used to display the contents of a file one screen at a time. This is useful for viewing large files like FASTQ files, where the data is too big to fit on a single screen.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> <span class="op">&lt;</span>path_file_to_display<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Navigation keys in more
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><strong>space</strong>: pressing the <em>space</em> key advances the display by one full screen, allowing you to scroll through the content quickly.</li>
<li><strong>enter</strong>: Ppressing <em>enter</em> moves the display down by one line. This is useful if you want to slowly scroll through the content line by line.</li>
<li><strong>q</strong>: pressing <em>q</em> will quit the more command and return you to the terminal prompt. You can use this key if you’re done viewing the file or don’t want to scroll through the entire file.</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> chrom22.fastq </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 1</strong></span> How many lines are used to represent each read?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 2</strong></span> How many sequences are present in <code>human_genome.fasta</code>? What do they correspond to? Use the command <code>grep</code> to display only the identifier lines.</p>
</div>
<p><strong>grep</strong></p>
<p>Grep, short for “global regular expression print”, is a command used for searching and matching text patterns in files contained in the regular expressions.</p>
<p>Display the lines with a pattern:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&lt;pattern&gt;"</span> <span class="op">&lt;</span>file_path<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command searches through the specified file and outputs the lines that contain the given pattern.</p>
<p>Count the lines with a pattern:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&lt;pattern&gt;"</span> <span class="op">&lt;</span>file_path<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The -c option tells grep to count the number of lines that match the specified pattern rather than displaying them. This is useful for quickly determining how many occurrences of the pattern exist in the file.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the identifier lines</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> reference/human_genome.fasta </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the identifier lines</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> reference/human_genome.fasta </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="fastq---quality-control" class="level2">
<h2 class="anchored" data-anchor-id="fastq---quality-control">FASTQ - Quality control</h2>
<p>Use the <code>fastqc</code> tool to assess the quality of the reads. Before running <code>fastqc</code>, create a directory to store the output reports (with <code>mkdir</code>command).</p>
<p><strong>FastQC</strong></p>
<p>FastQC is a quality control tool for high-throughput sequence data that provides an overview of various quality metrics, helping to identify potential issues with sequencing data.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> <span class="op">&lt;</span>output_directory<span class="op">&gt;</span> <span class="op">&lt;</span>fastq_file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The option <code>-o &lt;output_directory&gt;</code> specifies the directory where the FastQC output files will be saved.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> FastQCReport</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> FastQCReport chrom22.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>fastcq</code> analyzes the file <code>chrom22.fastq</code> and saves the results in the <code>FastQCReport</code> directory.</p>
</div>
</div>
</div>
<p>Move the HTML output files to your local computer to view the results in a browser. Use the <code>scp</code> command for this (in a local terminal).</p>
<p><strong>scp</strong></p>
<p>The scp command stands for “secure copy protocol.” It is used to securely transfer files between a local computer and a remote server or between two remote servers over an encrypted SSH connection. This ensures that the data is protected during transfer.</p>
<p>To copy file from a server to your computer, <strong>open a new terminal</strong> and use scp like this:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@<span class="op">&lt;</span>server_adress<span class="op">&gt;</span>:<span class="op">&lt;</span>source_file_path<span class="op">&gt;</span> <span class="op">&lt;</span>target_file_path<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
View the path of a file
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you not sure of the path .html file, use the command <code>realpath</code> to known it. <code>realpath &lt;file&gt;</code></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@tp.lbgi.fr:~/TdExome/FastQCReport/chrom22_fastqc.html .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>Note: Replace <code>&lt;login&gt;</code> with your username. The tilde <code>~</code> is a shortcut for your home directory <code>/home/&lt;username&gt;</code>.</em></p>
</div>
</div>
</div>
<p>Now that the HTML file is on your computer, you can access it like any other file using your file explorer.</p>
<p>In the terminal where you used the <code>scp</code> command, run the <code>pwd</code> command to check your current directory. This will help you locate the file in your file explorer.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3</strong></span> How many sequences are available?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4</strong></span> What is the sequence length?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5</strong></span> Do you find poor quality sequences?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 6</strong></span> What do you think of the overall data quality?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 7</strong></span> Which part of the reads exhibit a lower quality?</p>
</div>
</section>
<section id="alignement---mapping-to-the-reference-genome-fastq-to-bam" class="level2">
<h2 class="anchored" data-anchor-id="alignement---mapping-to-the-reference-genome-fastq-to-bam">Alignement - Mapping to the reference genome (FASTQ to BAM)</h2>
<section id="indexing-reference-file" class="level3">
<h3 class="anchored" data-anchor-id="indexing-reference-file">Indexing reference file</h3>
<p>Before we can align sequencing reads to a reference genome, we must prepare the reference file using two key tools: <code>bwa index</code> and <code>samtools faidx</code>.</p>
<p><strong>bwa index</strong>: The bwa index command is used to create an index of the reference genome (in this case, human_genome.fasta). This index allows the BWA algorithm to efficiently search and align sequencing reads to the reference. Without the index, BWA would not be able to quickly identify where in the genome each read belongs, making the alignment process slow or impossible. The index is essential for speeding up the read mapping process.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> index <span class="op">&lt;</span>reference_file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>samtools faidx</strong>: Similarly, samtools faidx generates an index for the reference FASTA file. This index allows samtools and other bioinformatics tools to quickly retrieve specific sequences or regions from the reference genome. This is useful when working with large genomes where extracting specific chromosomes or regions manually would be inefficient. The .fai file generated by this command is used by downstream tools for fast reference lookups.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx <span class="op">&lt;</span>reference_file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> reference</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> index human_genome.fasta</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx human_genome.fasta</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>These commands may take some time to complete (one hour). To save time, you can use the ln command to create links to already generated files. If you’ve already started the genome indexing process but interrupted it (with <code>ctrl + C</code>), make sure to remove any partially generated files before creating the links:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> reference/human_genome.fasta.<span class="pp">*</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /data/human/reference/human_genome.fasta.<span class="pp">*</span> reference/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>You get a lots of new files, forming a small database that you will use to align reads.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Know where you are
</div>
</div>
<div class="callout-body-container callout-body">
<p>To find out your current location within the directory structure, use the command <code>pwd</code> (stands for “print working directory”). It outputs the full path of the directory you are currently in, helping you confirm your current location in the file system.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="bu">pwd</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="alignment-using-bwa" class="level3">
<h3 class="anchored" data-anchor-id="alignment-using-bwa">Alignment using BWA</h3>
<p><strong>bwa</strong></p>
<p>Create a mapping directory (at the same level as <code>reference</code> and <code>FastQCReport</code> directories), move there and run BWA with the <strong>maximum exact matches (MEM)</strong> algorithm. This algorithm is faster and more accurate than previous BWA algorithms for aligning high-throughput sequencing data.</p>
<p>The command <code>bwa mem</code> uses BWA use the MEM algorithm:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem <span class="at">-o</span> <span class="op">&lt;</span>path_of_output<span class="op">&gt;</span> <span class="op">&lt;</span>path_of_reference<span class="op">&gt;</span> <span class="op">&lt;</span>path_of_reads<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Parameters:</p>
<ul>
<li><code>-o</code>: This option specifies the output file. In this case, the aligned reads will be saved in the mapping directory as chrom22.sam.</li>
<li><code>&lt;path_of_reference&gt;</code>: This is the reference genome file against which the reads will be aligned.</li>
<li><code>&lt;path_of_reads&gt;</code>: This is the input file containing the sequencing reads that you want to align to the reference genome.</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> mapping</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem <span class="at">-o</span> mapping/chrom22.sam reference/human_genome.fasta chrom22.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 8</strong></span> View the file built by BWA.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> mapping/chrom22.sam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9</strong></span> What is the format of this file? What information can be found in it?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 10</strong></span> In what order are the sequences presented?</p>
</div>
<p>To manipulate the mapping file, we’ll use&nbsp;<a href="http://www.htslib.org/">SAMtools</a>.</p>
<p><strong>samtools view</strong> This command is specifically used to convert between these formats and to filter, extract, or manipulate alignments.</p>
<p>To transform SAM file into BAM use the command like this:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-S</span> <span class="at">-b</span> <span class="at">-o</span> <span class="op">&lt;</span>name_of_BAM_output<span class="op">&gt;</span> <span class="op">&lt;</span>name_of_SAM_input<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>With:</p>
<ul>
<li><code>-S</code>: indicate that the input file is in SAM format.</li>
<li><code>-b</code>: tell samtools to output the result in BAM format.</li>
<li><code>-o</code>: specify the name of the output file.</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> mapping </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-S</span> <span class="at">-b</span> <span class="at">-o</span> chrom22.bam chrom22.sam </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 11</strong></span> What does the BAM file represent in relation to the SAM file?</p>
</div>
<p><strong>ls</strong></p>
<p>The <code>ls</code> command is used to list the contents of a directory in Linux. It provides a way to see files and subdirectories within the specified directory.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">[-</span><span class="ss">l</span><span class="pp">]</span> <span class="op">&lt;</span>directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Options:</p>
<ul>
<li><code>-l</code>: This option provides a detailed listing. In this format, additional information about each file or directory is displayed, including:
<ol type="1">
<li>File permissions</li>
<li>Number of links</li>
<li>Owner name</li>
<li>Owner group</li>
<li>File size (in bytes)</li>
<li>Last modified date and time</li>
<li>File or directory name</li>
</ol></li>
</ul>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 12</strong></span> Compare the size of the SAM and BAM files.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>The file now needs to be sorted and indexed with <code>samtools sort</code> and <code>samtools index</code>.</p>
<p><strong>samtools sort</strong></p>
<p>The samtools sort command is used to sort a BAM file. Sorting is essential for many downstream analyses, such as variant calling, as it organizes the reads in a way that allows for efficient processing. The command takes the following parameters:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> sort <span class="op">&lt;</span>name_of_BAM_file<span class="op">&gt;</span> -o <span class="op">&lt;</span>name_of_sorted_BAM_file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>samtools index</strong></p>
<p>The samtools index command is used to create an index file for a sorted BAM file. Indexing allows for fast random access to data in the BAM file, making it easier and quicker to retrieve specific regions of interest during analysis. The command takes the following parameter:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index <span class="op">&lt;</span>name_of_sorted_BAM_file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> sort chrom22.bam <span class="at">-o</span> chrom22.sorted.bam</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index chrom22.sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 13</strong></span> How is the file sorted now? To view its contents, use the command <code>samtools view</code>.</p>
</div>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="op">&lt;</span>file_name<span class="op">&gt;</span> <span class="kw">|</span> <span class="fu">more</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The <strong>pipe</strong> symbol (<code>|</code>) is used to connect the output of one command directly into the input of another. This allows for the chaining of commands, enabling you to perform complex tasks efficiently without the need to create intermediate files.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">command1</span> <span class="kw">|</span> <span class="ex">command2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pipe Shortcut
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Windows and Linux: <code>Alt Gr + 6</code></li>
<li>Mac keyboard: <code>Option + Maj + L</code></li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view chrom22.sorted.bam <span class="kw">|</span> <span class="fu">more</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>To view few statistics about the mapping, use idxStats:</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> idxstats chrom22.sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 14</strong></span> How do you explain these results?</p>
</div>
<p>Select the part of the alignment corresponding to chromosome 3. To filter the file, use again the command <code>samtools view</code> but add the region.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-b</span> <span class="at">-o</span> <span class="op">&lt;</span>name_of_BAM_output<span class="op">&gt;</span> <span class="op">&lt;</span>name_of_BAM_input<span class="op">&gt;</span> <span class="op">&lt;</span>region<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><code>-b</code>: This option indicates that the output should be in BAM format.</li>
<li><code>-o &lt;name_of_BAM_output&gt;</code>: This specifies the name of the output BAM file.</li>
<li><code>&lt;name_of_BAM_input&gt;</code>: This is the name of the input BAM file you want to filter.</li>
<li><code>&lt;region&gt;</code>: Specify the region you want to extract, such as <code>chr3</code> or <code>chr3:1-1000000</code> to limit the output to a specific range on chromosome 3.</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb36"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-b</span> <span class="at">-o</span> chrom22.sorted.filter.bam chrom22.sorted.bam chr22</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Then index this file with <code>samtools index</code> as previously.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index chrom22.sorted.filter.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Download the two 2 bam files of the alignment of chromosome 22 (the <code>.bam</code> and the <code>.bam.bai</code>).</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-24-contents" aria-controls="callout-24" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-24" class="callout-24-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb38"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>login<span class="op">&gt;</span>@tp.lbgi.fr:~/TdExome/mapping/chrom22.sorted.filter.bam<span class="pp">*</span> .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>Note: Replace <code>&lt;login&gt;</code> with your username. The tilde <code>~</code> is a shortcut for your home directory <code>/home/&lt;username&gt;</code>.</em></p>
</div>
</div>
</div>
</section>
<section id="visualization-of-the-bam-file" class="level3">
<h3 class="anchored" data-anchor-id="visualization-of-the-bam-file">Visualization of the BAM file</h3>
<p>BAM files can be visualized using a browser. Display the filtered BAM data using <a href="https://igv.org/app/">IGV</a> with Human hg38.</p>
<p>You can use your own files or used files available online.</p>
<ul>
<li>Load the bam file <code>chrom22.sorted.filter.bam</code> and <code>chrom22.sorted.filter.bam.bai</code> with Local (Tracks &gt; Local).</li>
<li>Load the bam file https://lbgi.fr/~merlat/TD3/chr22_filter.bam &amp; the index file https://lbgi.fr/~merlat/TD3/chr22_filter.bai with URL (Tracks &gt; URL).</li>
</ul>
<p>Zoom in to display a small region of chromosome 22 (or a gene such as APOL5). Configure the display to highlight mismatches.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 15</strong></span> Are reads exclusively mapped on exonic regions? Why?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 16</strong></span> Do you think that all observed mismatches correspond to variants?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 17</strong></span> What criteria have to be fulfilled to infer variants?</p>
</div>
</section>
</section>
<section id="calling-single-nucleotide-variations-snvs-pileup-interval-and-bed-files" class="level2">
<h2 class="anchored" data-anchor-id="calling-single-nucleotide-variations-snvs-pileup-interval-and-bed-files">Calling single nucleotide variations (SNVs) – pileup, interval and BED files</h2>
<section id="generation-of-a-pileup-file" class="level3">
<h3 class="anchored" data-anchor-id="generation-of-a-pileup-file">Generation of a Pileup file</h3>
<p>At the same level as <code>reference</code>, <code>FastQCReport</code> and <code>mapping</code> directories, create a directory <code>variantCalling</code>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb39"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> variantCalling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>A pileup is a column wise representation of the aligned read - at the base level - to the reference. The pileup file summarizes all data from the reads at each genomic region that is covered by at least one read.</p>
<p>To generate a pileup file using SAMtools, use <code>samtools mpileup</code> command with the specified options.</p>
<p><strong>samtools mpileup</strong></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> mpileup <span class="at">-f</span> <span class="op">&lt;</span>reference_file<span class="op">&gt;</span> <span class="op">&lt;</span>bam_file<span class="op">&gt;</span> <span class="op">&gt;</span> <span class="op">&lt;</span>pileup_output<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Parameters:</p>
<ul>
<li><code>-f &lt;reference_file&gt;</code>: Specifies the reference genome file to which the reads are aligned.</li>
<li><code>&lt;bam_file&gt;</code>: The input BAM file that contains the aligned reads. It must be sorted and indexed for efficient querying.</li>
<li><code>&gt;</code>: Redirects the standard output (stdout) to a file, saving the results of the command into the specified file rather than displaying them in the terminal.</li>
<li><code>&lt;pileup_output&gt;</code>: This writes the output to the specified pileup file.</li>
</ul>
<p>Generate the pileup file and adjust the mapping quality to 50.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-26-contents" aria-controls="callout-26" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-26" class="callout-26-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb41"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> mpileup <span class="at">-f</span> reference/human_genome.fasta <span class="at">-C</span> 50 aligned_reads.bam <span class="op">&gt;</span> variantCalling/chrom22.pileup</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Search online for the significance of each column.</p>
</section>
<section id="filtering-pileup-to-get-a-list-of-snvs" class="level3">
<h3 class="anchored" data-anchor-id="filtering-pileup-to-get-a-list-of-snvs">Filtering pileup to get a list of SNVs</h3>
<p>The pileup described all aligned positions. To call SNVs, we have to filter:</p>
<ul>
<li>Aligned positions that differ from the reference base</li>
<li>Positions covered by minimum amount of reads (e.g.&nbsp;at least 10 reads)</li>
<li>Sites with variants with sufficient quality base scores</li>
</ul>
<section id="first-filter-depth-of-coverage-and-read-base-quality" class="level4">
<h4 class="anchored" data-anchor-id="first-filter-depth-of-coverage-and-read-base-quality">First filter: depth of coverage and read base quality</h4>
<p>This command generates an interval format file.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">perl</span> /bin/pileup_parser.pl <span class="op">&lt;</span>input_pileup<span class="op">&gt;</span> <span class="op">&lt;</span>ref_base_column<span class="op">&gt;</span> <span class="op">&lt;</span>read_bases_column<span class="op">&gt;</span> <span class="op">&lt;</span>base_quality_column<span class="op">&gt;</span> <span class="op">&lt;</span>coverage column<span class="op">&gt;</span> <span class="op">&lt;</span>qv cutoff<span class="op">&gt;</span> <span class="op">&lt;</span>coverage cutoff<span class="op">&gt;</span> <span class="op">&lt;</span>SNPs only<span class="pp">?</span><span class="op">&gt;</span> <span class="op">&lt;</span>output bed<span class="pp">?</span><span class="op">&gt;</span> <span class="op">&lt;</span>coord_column<span class="op">&gt;</span> <span class="op">&lt;</span>out_file<span class="op">&gt;</span> <span class="op">&lt;</span>total_diff<span class="op">&gt;</span> <span class="op">&lt;</span>print_qual_bases<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Arguments:</p>
<ul>
<li><code>&lt;input_pileup&gt;</code>: Path to your input pileup file.</li>
<li><code>&lt;ref_base_column&gt;</code>: The column index (starting from 1) of the reference nucleotide.</li>
<li><code>&lt;read_bases_column&gt;</code>: The column index of the bases at that position from aligned reads.</li>
<li><code>&lt;base_quality_column&gt;</code>: The column index of the Phred quality.</li>
<li><code>&lt;coverage_column&gt;</code>: The column index of the depth of coverage.</li>
<li><code>&lt;qv_cutoff&gt;</code>: Quality value cutoff.</li>
<li><code>&lt;coverage_cutoff&gt;</code>: Coverage cutoff.</li>
<li><code>&lt;SNPs_only&gt;</code>: If you only want SNPs (yes/no).</li>
<li><code>&lt;output_bed&gt;</code>: If you want to output in BED format (yes/no).</li>
<li><code>&lt;coord_column&gt;</code>: The coordinate system.</li>
<li><code>&lt;out_file&gt;</code>: Path where you want to save the output (e.g., output.txt).</li>
<li><code>&lt;total_diff&gt;</code>: If you want to print the total number of differences (yes/no).</li>
<li><code>&lt;print_qual_bases&gt;</code>: If you want to print the quality bases (yes/no).</li>
</ul>
<p>Use this command to parse the pileup file by filling with the correct columns and the folowing options:</p>
<ul>
<li>Do not consider read bases with quality lower than = <strong>20</strong>.</li>
<li>Do not report positions with coverage lower than = <strong>10</strong>.</li>
<li>Only SNPs.</li>
<li>Do not produce a output in BED format.</li>
<li>Use the coordinate system 2.</li>
<li>Do not print the total number of difference.</li>
<li>Print the quality bases.</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb43"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> variantCalling/</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="ex">/bin/pileup_parser.pl</span> chrom22.pileup 3 5 6 4 20 10 Yes No 2 chrom22.pileup.parsed No Yes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 18</strong></span> How many regions do you obtain? Use the command <code>wc -l</code> to count the number of lines.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-28-contents" aria-controls="callout-28" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-28" class="callout-28-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> chrom22.pileup.parsed </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="second-filter-snv-quality-50" class="level4">
<h4 class="anchored" data-anchor-id="second-filter-snv-quality-50">Second filter: SNV quality (&gt;50)</h4>
<p>In the interval file, SNV quality is indicated in column 7. Use awk to keep only lines where column 7 &gt; 50, combine with wc -l to count the number of keeped lines.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 19</strong></span> How many reliable SNVs have been called? Use awk to keep only lines where column 7 &gt; 50, combine with wc -l to count the number of keeped lines.</p>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-29-contents" aria-controls="callout-29" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-29" class="callout-29-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb45"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'$7 &gt; 50'</span>  chrom22.pileup.parsed2 <span class="kw">|</span><span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="annotation-of-variants" class="level2">
<h2 class="anchored" data-anchor-id="annotation-of-variants">Annotation of variants</h2>
<p>Go to Variant Effect Predictor : http://grch37.ensembl.org/Homo_sapiens/Tools/VEP</p>
<p>Click on Ensembl default below the white frame for input data</p>
<p>Paste the URL : http://lbgi.fr/~weber/GC/TD3/2_VCF/chr22_filter.vcf.gz corresponding to the chr22 VCF file in the URL field.</p>
<p>Select the following options :</p>
<ul>
<li>Transcript database to use</li>
<li>RefSeq transcripts</li>
<li>Additional configuration</li>
<li>Variants and frequency data</li>
<li>Frequency data for co-located variants:</li>
<li><strong>DISABLE</strong> 1000 Genomes global minor allele frequency</li>
<li><strong>ENABLE</strong> gnomAD (exomes) allele frequencies</li>
<li>Predictions</li>
<li>Pathogenicity predictions</li>
<li><strong>ENABLE</strong> CADD</li>
<li>Filtering options</li>
<li>Filters</li>
<li>Restrict results: Show one selected consequence per gene</li>
</ul>
<p>Then RUN !</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 20</strong></span> How many missense variants are there in this file ? (approximately)</p>
</div>
<p>Create a filter to select only missense_variant (Consequence is missense_variant)</p>
<p>Focus on 22:23482483-23482483.</p>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 21</strong></span> On which gene is the variant ? Which exon ?</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 22</strong></span> Why is this variant difficult to interpret ? (MAF, scores)</p>
</div>
<div id="exr-1" class="custom theorem exercise">
<p><span class="theorem-title"><strong>Exercise 23</strong></span> Bonus : is the variant referenced in ClinVar ?</p>
</div>


</section>
</section>

</main> <!-- /main -->
<script>
    var custom_title = document.querySelectorAll('.custom .theorem-title');
    
    for (let i = 0; i < custom_title.length; i++ ) {
       var mod_name = custom_title[i].innerHTML;
       custom_title[i].innerHTML = mod_name.replace("Exercise", "Question");
    };
    </script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Copyright 2024, Dorine Merlat. Built with Quarto.</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>